### ISSUE TITLE
Gene-drop estimation of genotype probabilities

### ISSUE THREAD

**Issue by timothymillar** [pystatgen/sgkit#1107]
A gene-drop simulation treats a pedigree as a stochastic model from which alleles/genotypes are drawn at random following (simplified) Mendelian inheritance. These simulations typically don't model linkage among loci and are used to approximate expected genotype distributions per locus.

I already have a rough implementation in numba, however there's multiple ways in which the results could be reported. The most general way to store the results of a simulation is as a call-genotype array per replicate, but this can result in a massive array (replicates * variants * samples * ploidy). It's much more efficient to summaries the replicates on the fly and there are a few options here. An obvious choice is to store the distribution of possible genotypes for each call (variants * samples * genotypes). This captures most of the information about individual calls. We could also look at reporting the probability of the observed call genotype which would be simpler (variants * samples).

**Comment by jeromekelleher** [comment]
How about returning the replicates independently as a generator, like [msprime does](https://tskit.dev/msprime/docs/latest/replication.html#running-replicate-simulations)? Summarising the stream of results should be something that could be done in client code then (if we give some useful examples)?

**Comment by jeromekelleher** [comment]
It would be reasonable to treat each replicate as an independant thing though, wouldn't it?  Surely each replicate (a variant dataset) could be analysed by itself, and then what you're doing is providing a way to aggregate stats over multiple datasets?

I think trying to run the simulations *and* do the full analysis of each replicate in a single function call would be too much. You need to separate the process of simulation from analysis, or it'll be a mess.

**Comment by timothymillar** [comment]
> I think trying to run the simulations and do the full analysis of each replicate in a single function call would be too much

Agreed. I was fixated on returning a single dataset which made me think it would be unwieldy to return every iteration.

> Surely each replicate (a variant dataset)

That makes much more sense, I thought you were suggesting a generator yielding arrays rather than full datasets. Returning a dataset per replicate would certainly be the most flexible approach (assuming the performance is adequate).

Would it be more consistent to return single replicate rather than a generator? Multiple replicates could then be achieved with a for loop or a generator expression.

**Comment by timothymillar** [comment]
Ah, or we could follow msprime and only return a generator if `num_replicates` is specified. I'm slowly catching up 😄

**Comment by jeromekelleher** [comment]
A single replicate is a good place to start, let's see how that works out. Can add support for a generator over multiple replicates later if it seems helpful
